source/slang/slang-ir-link.cpp

```diff
@@ -1519,6 +1520,146 @@ static void diagnoseUnresolvedSymbols(TargetRequest* req, DiagnosticSink* sink,
+    {
+        //I don't need a sorted list, just the max so I can make an array of uint[(maxOffset/4)+1]
+        int64_t maxOffset = 0;
+        for (auto& i : bindingToInstList.second)
+        {
+            int64_t currOffset = int64_t(i->findDecoration<IRGLSLOffsetDecoration>()->getOffset()->getValue());
+            maxOffset = (maxOffset < currOffset) ? currOffset : maxOffset;
+        }
+        auto instToSwitch = *bindingToInstList.second.begin();
+        builder.setInsertBefore(instToSwitch);
+        
+        // now we know the max offset and can construct a new storageBuffer with key of type maxOffset/4 (32bit uint per atomic_uint)
+        auto elementType = builder.getArrayType(
+            builder.getUIntType(),
+            builder.getIntValue(builder.getUIntType(), (maxOffset / sizeof(uint32_t))+1)
+        );
+
+        // now we make our struct which goes within with some name (does not matter), 
+        StringBuilder nameStruct;
+        nameStruct << "atomic_uints";
+        nameStruct << bindingToInstList.first;
+        nameStruct << "_t";
+        nameStruct << "_paramGroup";
+        auto structType = builder.createStructType();
+        builder.addNameHintDecoration(structType, nameStruct.produceString().getUnownedSlice());
+
+        // our "_data" struct member
+        auto elementBufferKey = builder.createStructKey();
+        builder.addNameHintDecoration(elementBufferKey, UnownedStringSlice("_data"));
+        auto elementBufferType = elementType;
+        auto _dataField = builder.createStructField(structType, elementBufferKey, elementBufferType);
+        
+        // make our valid storage buffer type
+        auto std430 = builder._createInst(sizeof(IRTypeLayoutRules), builder.getType(kIROp_Std430BufferLayoutType), kIROp_Std430BufferLayoutType);
+        IRGLSLShaderStorageBufferType* storageBuffer;
+        {
+            IRInst* ops[] = { structType, std430 };
+            storageBuffer = builder.createGLSLShaderStorableBufferType(2, ops);
+        }
+
+        // link storage buffer to any known atomic_uint
+        instToSwitch->setFullType(storageBuffer);
+        
+        // now we need to make all call's to:
+        // 1. convert all atomic_uint function type parameters to a inlined call of atomic_uint->inout int ref of 
+        // field->element-addr based on location(offset) from the atomic_uint->storage_buffer logic above
+        for (auto& i : bindingToInstList.second)
+        {
+            int64_t currOffset = int64_t(i->findDecoration<IRGLSLOffsetDecoration>()->getOffset()->getValue());
+            // we manage the "next" ptr with a seperate list instance since we don't want to add uses
+            // of an unrelated Call to our elementAddr
+            IRUse* next = nullptr;
+            for (auto use = i->firstUse; use; use = next)
+            {
+                next = use->nextUse;
+                auto user = use->user;
```

Comment:
Please explain why you are making a temporary variable unnecessarily.

Reply:
since we are using `replaceUsesWith`, emitting types, and setting types/operands, we are adding IRUses to an eventual parent atomic_uint->storageBuffer. If we do not store the old node we will be changing our IRUse list and be losing the `->next` node. As a result our code will not mangle all atomic_uint references and will stop after the first.

Reply:
this can be seen with the code emitted (or vulkan layers complaining that we have some 'ptr' to an int

I think there was a second reason I did this, but it has been a week and I don't remember; I do recall it being crash related though (node1->node2->node3 became node3->node1->node2, and then node2->node3->node1, causing an infinite loop)